Parser for PEDSnet ETL conventions document

l-k- opened the pull request:
Please let me know if this is what you had in mind for the ETL conventions doc parser.

l-k- committed:
Parser for PEDSnet ETL conventions document

bruth commented:
We follow a style guide to ensure there is consistency across our code bases. For Python we just adhere to PEP8. Please install [flake8](https://pypi.python.org/pypi/flake8) and run it against the file. It will output a bunch of things to fix. For Vim integration I recommend using [syntastic](https://github.com/scrooloose/syntastic) which works with a bunch of languages and errors are shown when you save your time. There is also a dedicated [flake8](https://github.com/nvie/vim-flake8) plugin.

l-k- commented:
done, thanks for the suggestion

bruth commented on ETLParser.py:
```diff
@@ -0,0 +1,166 @@
+                break
+
+        # Everything up to the first h2 element is model-level content.
+        content = ''
```
Not that it would have a great effect here, but as a general performance consideration, it is more efficient to append strings to a list then join them at the end.

l-k- replied:
didn't think of that, thanks

bruth commented on ETLParser.py:
```diff
@@ -0,0 +1,166 @@
+                break
+            content = content + line
+
+        self.add_content(content.strip('\n'))
```
By default `str.strip` will strip all whitespace from a string, so there is no need to specify `\n` (in fact you may miss carriage returns, spaces, and tabs).

bruth commented on ETLParser.py:
```diff
@@ -0,0 +1,166 @@
+        self.dict['tables'][table]['fields'][field]['etl_conventions'] = conv
+
+    def __str__(self):
+        return json.dumps(self.dict, indent=4)
```
Since this may not be the only output representation for the document, it would make more sense to move this logic down into `main` rather than assuming it to be the string representation of the document.

bruth commented on ETLParser.py:
```diff
@@ -0,0 +1,166 @@
+
+
+class Document():
+    def __init__(self, path):
```
For flexibility, this should take a file-like object (could be a file handler or a stream) and force the caller to handle closing the file or stream. This also enables you to consume the file line-by-line rather loading all the data into memory via `f.readlines()`. Any place you are consuming lines would just do:

``` python
for line in self.fileobj:
    ...
```

l-k- replied:
that does look a whole lot better, thanks for the suggestion

bruth commented on ETLParser.py:
```diff
@@ -0,0 +1,166 @@
+        self.dict['tables'][table]['content'] = content
+
+    def add_field(self, table, field,
+                  required, data_type, desc, conv):
```
Only wrap arguments if the line exceeds 79 characters. I don't believe this one does.

bruth commented on ETLParser.py:
```diff
@@ -0,0 +1,166 @@
+        # Skip everything up to and including the h1 element (the title).
+        while self.pos < len(self.lines):
+            line = self.lines[self.pos]
+            self.pos = self.pos + 1
```
`self.pos += 1` is preferred to the expanded version.

bruth replied:
That being said, keeping track of the position should not longer be needed why using the `for` loop approach mentioned above.

bruth replied:
To complete the thought, the `line` itself would be cached on the document, not the position in the list of lines.

bruth commented on ETLParser.py:
```diff
@@ -0,0 +1,166 @@
+            if _h2.match(line):
+                break
+
+            field = re.match(r'^(?P<name>[^\|]*)\|(?P<req>[^\|]*)\|'
```
Move this regexp outside of the class since it will be reused multiple times.

bruth replied:
As well as an other regexes defined inline.

bruth commented on ETLParser.py:
```diff
@@ -0,0 +1,166 @@
+                desc = field.group('desc').strip()
+                conv = field.group('conv')
+                if conv:
+                    conv = conv.strip()
```
No need for the condition. `conv = field.group('conv').strip()` should be fine.

l-k- replied:
if ETL conventions aren't specified, then the field is shortened to 'Field |Required | Data Type | Description', usually without a training '|', and so field.group('conv') ends up being not set. 
So I get the error 
'NoneType' object has no attribute 'strip'
unless I do check for this.

bruth replied:
Ok that makes sense.

bruth replied:
I missed the `?` in the regex for `conv`!

bruth commented on ETLParser.py:
```diff
@@ -0,0 +1,166 @@
+
+if __name__ == '__main__':
+    if len(sys.argv) != 2:
+        sys.exit('Usage: %s %s' % (sys.argv[0], 'full_path_to_markdown_file'))
```
To support taking a file or stream, the path to the file should be optional. If not provided, then default to `sys.stdin` as the file-like object.

``` python
if len(sys.argv) == 1:
    main(sys.stdin)
    return

path = sys.argv[1]

# This handles checking if the file exists, so there is no need to do
# it explicitly. Also then ensures the file is properly closed once the
# program exits.
with open(path) as f:
    main(f)
```

l-k- commented:
Thanks for the review! I addressed your comments. Let me know if you want the commit squashed in right away -- I left it in case you want to see what I just changed.

bruth commented on ETLParser.py:
```diff
@@ -0,0 +1,170 @@
+
+class Document():
+    def __init__(self, fileobj):
+        self.dict = {}
```
Rename `dict` to `model` to clarify what the top-level value represents.

bruth commented on ETLParser.py:
```diff
@@ -0,0 +1,170 @@
+    def __init__(self, fileobj):
+        self.dict = {}
+        self.dict['content'] = ''
+        self.dict['tables'] = {}
```
Clean up all of the dict initializations to reduce the number of lookups in the dictionary itself and make it more readable.

``` python
self.model = {
    'content': '',
    'tables': {},
}
```

bruth commented on ETLParser.py:
```diff
@@ -0,0 +1,170 @@
+        if self.line == '':
+            return False
+
+        # Table name line looks something like this: "## 1.1 PERSON".
```
The `_h2` regex could be updated to match on this structure `^\#\# \d+\.\d+` so other h2s would not need to be considered.

bruth replied:
I guess the `_table_name` does already this. Is there a need to match on `h2` otherwise?

bruth commented:
Also let's keep the parsers in the same directory. Rename it to `parsers/etl_conventions.py`. Out of curiosity are you using Python 2 or 3 for this? It should work on both, but I prefer all new Python things to be developed against Python 3 especially if they are self-contained scripts.

l-k- commented:
Addressed the comments, thanks. I was indeed using Python 2, but updated the script to work with Python 3 now.

bruth commented on parsers/etl_conventions.py:
```diff
@@ -0,0 +1,165 @@
+            field = _get_field(line)
+
+            if field:
+                self.add_field(table_name, *field[:])
```
No need to copy the list when it is being unpacked.

bruth commented on parsers/etl_conventions.py:
```diff
@@ -0,0 +1,165 @@
+        # Everything up to the first table is model-level content.
+        self.line = ''
+        content = []
+        for line in self.fileobj:
```
Add empty line above for readability.

bruth commented on parsers/etl_conventions.py:
```diff
@@ -0,0 +1,165 @@
+
+        # Read the rest of the table contents, whatever comes after
+        # the fields specification table, up to the next table.
+
```
Remove empty between comment and block it pertains to.

bruth commented on parsers/etl_conventions.py:
```diff
@@ -0,0 +1,165 @@
+        self.get_content()
+
+        while self.get_next_table():
+            pass
```
These two methods are the entrypoint for the parser. Rather than calling them in the constructor and calling `get` later, let's combine them into a `parse` method:

``` python
def parse(self):
    self.get_content()

    while self.get_next_table():
        pass

    return self.model
```

The caller can then initialize the document and call `parse` when ready.

bruth commented:
Looks good, for the final bit, can you include a copy of the ETL conventions document to test.

```
#!/bin/bash

python3 etl_conventions.py test.md | diff expected.json - > /dev/null

# Message stating that the output did not match.
if [ $? > 0 ]; then
    echo 'Output did not match'
fi
```

For this to work, update the `json.dumps` call to include `sort_keys=True`.

bruth commented:
Looking good. After looking the through the output, can you handle [this case](https://github.com/l-k-/pedsnet-origins/blob/master/parsers/tests/expected.json#L181) (the regex would match everything up to the symbol) and also replace the spaces in table names with underscores. Also a minor nit to move the `signal` import statement below `import sys`.

bruth commented:
Looks great. Please squash the commits and I will merge!

l-k- commented:
Squashed. Thanks for the reviews!
